File: src/main/java/parse/audio/Recorder.java
package parse.audio;

import io.github.jonelo.tts.engines.exceptions.SpeechEngineCreationException;
import util.audio.NativeTTS;

import javax.sound.sampled.*;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;

class Recorder extends Thread {
  private TargetDataLine micDataLine = null;
  private boolean stop = false;
  private ArrayList<Short> pcmBuffer = null;

  public Recorder(int audioDeviceIndex) throws IOException, SpeechEngineCreationException {
    AudioFormat format = new AudioFormat(16000f, 16, 1, true, false);
    DataLine.Info dataLineInfo = new DataLine.Info(TargetDataLine.class, format);
    TargetDataLine dataLine;
    try {
      dataLine = getAudioDevice(audioDeviceIndex, dataLineInfo);
      dataLine.open(format);
    } catch (LineUnavailableException e) {
      NativeTTS.tts("There is no available microphone.");
      System.exit(1);
      return;
    }

    this.micDataLine = dataLine;
    this.stop = false;
    this.pcmBuffer = new ArrayList<>();
  }

  private static TargetDataLine getDefaultCaptureDevice(DataLine.Info dataLineInfo)
          throws LineUnavailableException {
    if (!AudioSystem.isLineSupported(dataLineInfo)) {
      throw new LineUnavailableException(
              "Default capture device does not support the audio format required by Picovoice (16kHz, 16-bit, linearly-encoded, single-channel PCM).");
    }

    return (TargetDataLine) AudioSystem.getLine(dataLineInfo);
  }

  private static TargetDataLine getAudioDevice(int deviceIndex, DataLine.Info dataLineInfo)
          throws LineUnavailableException {
    if (deviceIndex >= 0) {
      try {
        Mixer.Info mixerInfo = AudioSystem.getMixerInfo()[deviceIndex];
        Mixer mixer = AudioSystem.getMixer(mixerInfo);

        if (mixer.isLineSupported(dataLineInfo)) {
          return (TargetDataLine) mixer.getLine(dataLineInfo);
        }
      } catch (LineUnavailableException e) {
        e.printStackTrace();
      }
    }

    // use default capture device if we couldn't get the one requested
    return getDefaultCaptureDevice(dataLineInfo);
  }

  @Override
  public void run() {
    micDataLine.start();

    ByteBuffer captureBuffer = ByteBuffer.allocate(512);
    captureBuffer.order(ByteOrder.LITTLE_ENDIAN);
    short[] shortBuffer = new short[256];

    while (!stop) {
      micDataLine.read(captureBuffer.array(), 0, captureBuffer.capacity());
      captureBuffer.asShortBuffer().get(shortBuffer);
      for (short value : shortBuffer) {
        this.pcmBuffer.add(value);
      }
    }
  }

  public void end() {
    this.stop = true;
    // Close the micDataLine to release resources
    if (micDataLine != null) {
      micDataLine.close();
    }
  }

  public short[] getPCM() {
    short[] pcm = new short[this.pcmBuffer.size()];
    for (int i = 0; i < this.pcmBuffer.size(); ++i) {
      pcm[i] = this.pcmBuffer.get(i);
    }
    return pcm;
  }
}

File: src/main/kotlin/util/audio/Recorder.kt
package util.audio

import java.nio.ByteBuffer
import java.nio.ByteOrder
import javax.sound.sampled.AudioFormat
import javax.sound.sampled.AudioSystem
import javax.sound.sampled.DataLine
import javax.sound.sampled.LineUnavailableException
import javax.sound.sampled.TargetDataLine
import kotlin.system.exitProcess

class Recorder(
    audioDeviceIndex: Int,
) : Thread() {
    private var micDataLine: TargetDataLine? = null
    private val pcmBuffer = mutableListOf<Short>()
    private var stop = false

    init {
        val format = AudioFormat(16000f, 16, 1, true, false)
        val dataLineInfo = DataLine.Info(TargetDataLine::class.java, format)

        try {
            micDataLine = getAudioDevice(audioDeviceIndex, dataLineInfo)?.apply { open(format) }
                ?: throw LineUnavailableException("Failed to initialize audio device")
        } catch (_: LineUnavailableException) {
            NativeTTS.tts("There is no available microphone.")
            exitProcess(1)
        }
    }

    private fun getDefaultCaptureDevice(dataLineInfo: DataLine.Info): TargetDataLine? =
        if (AudioSystem.isLineSupported(dataLineInfo)) {
            AudioSystem.getLine(dataLineInfo) as? TargetDataLine
        } else {
            null
        }

    private fun getAudioDevice(
        deviceIndex: Int,
        dataLineInfo: DataLine.Info,
    ): TargetDataLine? =
        if (deviceIndex >= 0) {
            try {
                val mixerInfo = AudioSystem.getMixerInfo()[deviceIndex]
                val mixer = AudioSystem.getMixer(mixerInfo)
                if (mixer.isLineSupported(dataLineInfo)) {
                    mixer.getLine(dataLineInfo) as? TargetDataLine
                } else {
                    null
                }
            } catch (e: LineUnavailableException) {
                e.printStackTrace()
                null
            }
        } else {
            getDefaultCaptureDevice(dataLineInfo)
        }

    override fun run() {
        micDataLine?.let { line ->
            line.start()

            val captureBuffer = ByteBuffer.allocate(512).apply { order(ByteOrder.LITTLE_ENDIAN) }
            val shortBuffer = ShortArray(256)

            while (!stop) {
                line.read(captureBuffer.array(), 0, captureBuffer.capacity())
                captureBuffer.asShortBuffer()[shortBuffer]
                pcmBuffer.addAll(shortBuffer.toList())
            }
        }
    }

    fun end() {
        stop = true
        micDataLine?.close()
    }

    val pcm: ShortArray
        get() = ShortArray(pcmBuffer.size) { index -> pcmBuffer[index] }
}
